packages/neo-one-smart-contract-lib/src/ownership/Ownable.ts

```diff
@@ -0,0 +1,57 @@
+import { Address, SmartContract, createEventNotifier } from '@neo-one/smart-contract';
+
+const notifyTransferOwnership = createEventNotifier<Address | undefined, Address>('transfer_ownership', 'from', 'to');
+
+export function Ownable<TBase extends Constructor<SmartContract>>(base: TBase) {
+  abstract class OwnableClass extends base {
+    protected abstract initialOwner: Address;
+    private mutableOwner: Address | undefined = undefined;
+    private mutableOwnerInitialized = false;
+
+    public get owner(): Address | undefined {
+      return this.mutableOwner !== undefined ? this.mutableOwner : this.initializeOwner();
+    }
+
+    public renounceOwnership() {
+      this.onlyOwner();
+
+      this.mutableOwner = undefined;
+    }
+
+    public transferOwnership(to: Address): boolean {
+      this.onlyOwner();
+
+      notifyTransferOwnership(this.mutableOwner, to);
+      this.mutableOwner = to;
+
+      return true;
+    }
+
+    protected ownerOrThrow(): Address {
+      const owner = this.owner;
+
+      if (owner === undefined) {
+        throw new Error('no owner');
+      }
+
+      return owner;
+    }
+
+    protected onlyOwner() {
+      if (this.mutableOwner !== undefined && !Address.isCaller(this.mutableOwner)) {
```

Comment:
This should be using `this.ownerOrThrow` no? Seems like a bug - if there's no owner then this should throw I think.

Reply:
"bug" might be a little harsh, I'd say mine is between 4 and 12 operations more efficient and achieves same goal.

Reply:
I think he is suggesting something like:
```
protected onlyOwner() {
  if (!Address.isCaller(this.ownerOrThrow())) {
    throw new Error('not owner');
  }
}
```
(edit because assignment)

Reply:
(because if `this.mutableOwner` is undefined then this will pass always)

Reply:
@danwbyrne  yes, that was my glitch the !==undefined.  But after standup, we are favoring code readability over scrutinizing nominal amounts of gas for extra stack operations.   Change will be made, thank you!